protocol/contracts/governance/ElderElection.sol

```diff
@@ -134,8 +138,55 @@ contract ElderElection is EIP712, AccessControl {
     error NotCandidate(uint256 discordId);
     error TooManyEndorsements();
 
-    error DeadlineExpired(uint256 deadline);
+    error DeadlineExpired(uint256 lateBy);
+    error InvalidNonce(address account);
     error InvalidSignature(address account);
+    
+    struct EIP712Domain {
+        string name;
+        string version;
+        uint256 chainId;
+    }
+
+    bytes32 constant EIP712DOMAIN_TYPEHASH = keccak256("EIP712Domain(string name,string version,uint256 chainId)");
+
+    function hash(EIP712Domain memory _input) public pure returns (bytes32) {    
+        return keccak256(abi.encode(
+            EIP712DOMAIN_TYPEHASH,
+            keccak256(bytes(_input.name)),
+            keccak256(bytes(_input.version)),
+            _input.chainId
+        ));
+    }
+
+    struct EndorsementReq {
+        address account;
+        uint256[] discordIds;
+        uint256 deadline;
+        uint256 nonce;
+    }
+
+    bytes32 constant ENDORSEMENTREQ_TYPEHASH = keccak256("EndorsementReq(address account,uint256[] discordIds,uint256 deadline,uint256 nonce)");
+
+    function hash(EndorsementReq memory _input) public pure returns (bytes32) {
+        return keccak256(abi.encode(
+            ENDORSEMENTREQ_TYPEHASH,
+            _input.account,
+            hash(_input.discordIds),
+            _input.deadline,
+            _input.nonce
+        ));
+    }
+
+    function hash(uint256[] memory _input) public pure returns (bytes32) {
+        bytes memory encoded;
+        for (uint i = 0; i < _input.length; i++) {
```

Comment:
> Can abi.encodePacked(_input) not be used because it adds extra packing to arrays?

I originally had it as `abi.encode()` which doesn't work because the result includes the size. And examples I found were written as the above.

 I assumed (incorrectly) that `abi.encodePacked()` would still store the size. It turns out this is not the case, and, testing shows that `abi.encodePacked()` works here.